### PR TITLE
OpenZFS - 6363 Add UNMAP/TRIM functionality (v2)

### PR DESCRIPTION
**Pull request by tuxoko**
From #5925

Rebased from @dweeezil 's [ntrim2-next](https://github.com/dweeezil/zfs/tree/ntrim2-next)
Added trim related manpage changes.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by kpande** [comment]
er, this replaces #5925? why couldn't that PR just be updated instead?

**Comment by behlendorf** [comment]
@kpande, @tuxoko was nice enough to offer to help address some of the remaining issues with TRIM.  Thanks!  Step one is getting all the old and new tests passing.

> This happens a lot in ztest.
> WARNING: Pool 'ztest' has encountered an uncorrectable I/O failure and has been suspended.

This looks like one of those remaining issues which needs to be understood and fixed.  In the master branch we aren't seeing anything like this so it's likely TRIM related. 

> Also manualtrim_002_pos failed with
> TRIM interrupted it was expected to complete.

It also looks like several of the ZTS runs resulted in kernel panics.  So there's definitely some work to do root causing these issues.

File: module/zfs/vdev_queue.c [hunk 1]
```diff
@@ -736,7 +757,11 @@ vdev_queue_io_to_issue(vdev_queue_t *vq)
 	}
 
 	vdev_queue_pending_add(vq, zio);
-	vq->vq_last_offset = zio->io_offset + zio->io_size;
+
+	/* trim I/Os have no single meaningful offset */
+	if (zio->io_priority != ZIO_PRIORITY_AUTO_TRIM ||
+	    zio->io_priority != ZIO_PRIORITY_MAN_TRIM)
```
[the remark this answers is not in the document]
Comment (tuxoko) [reply module/zfs/vdev_queue.c]: Right, `||` doesn't make sense.

**Comment by KernelKyupCom** [comment]
Hi there,
I am also experiencing the "WARNING: Pool 'ztest3' has encountered an uncorrectable I/O failure and has been suspended." and the txg_sync_thread is blocked for more than 120 seconds (a hung task).

After some investigation it turns out there is ENOSPC returned by metaslab_alloc.
The actual call stack is:
#10 [ffffc90003c17bc0] metaslab_alloc at ffffffffa0508fdb [zfs]
#11 [ffffc90003c17cb0] zio_dva_allocate at ffffffffa058f470 [zfs]
#12 [ffffc90003c17da8] zio_execute at ffffffffa058b28a [zfs]
#13 [ffffc90003c17dd8] taskq_thread at ffffffffa04113d9 [spl]
#14 [ffffc90003c17f00] kthread at ffffffff8107ae2d
#15 [ffffc90003c17f50] ret_from_fork at ffffffff818001fa

I am creating zfs pool in a sparse file image.

I can easily reproduce it by copying two sets of small files - the linux kernel tree (including its git) and my ccache directory.
Basically I am importing the pool, copying the first set, creating snapshot1, deleting the files and then trimming. Then the same for the second set of files in another snapshot2. Exporting the pool and reimporting it, destroying the both snapshots, then again trim and export.
I am attaching my test script (just edit the paths): [zfs_test.sh.txt](https://github.com/zfsonlinux/zfs/files/1988313/zfs_test.sh.txt)
About the hung task it seems we are AB-BAdeadlocking. See the attached log: 
[lockdep_4.16.7.log](https://github.com/zfsonlinux/zfs/files/1988329/lockdep_4.16.7.log)

I hope this helps.
Any ideas? Some other useful data to capture to help resolving that?

Best regards,
Angel

File: module/zfs/metaslab.c [hunk 2]
```diff
@@ -3442,17 +3473,474 @@ metaslab_check_free(spa_t *spa, const blkptr_t *bp)
...
+
+	ASSERT(msp != NULL);
+	ASSERT(msp->ms_trimming_ts != NULL);
+	held = MUTEX_HELD(&msp->ms_lock);
+	if (!held)
+		mutex_enter(&msp->ms_lock);
+	VERIFY(!msp->ms_condensing);
+	if (msp->ms_loaded) {
+		range_tree_walk(msp->ms_trimming_ts, range_tree_add,
+		    msp->ms_tree);
```
Comment (aerusso) [inline module/zfs/metaslab.c]: Sorry if this has already been addressed elsewhere. During the time the trimming was being done (initiated by `metaslab_exec_trim`), `ms_trimming_ts` is removed from `msp->ms_tree`. If an allocation request occurs on such an `mg`, isn't it possible that `mg->mg_no_free_space` could be set (because of the unavailable range---the range being trimmed)?

And if so, is there code that would trigger a re-evaluation of the validity of that `mg_no_free_space` flag (now that said extent is again available for allocation)? If not, that could explain @KernelKyupCom's `ENOSPACE` issue.

**Comment by KernelKyupCom** [comment]
Hi guys,
I've been running for a month the code rebased to zfs-0.7.9-11. I haven't experienced the enospc issue anymore, but we hit another problem: in a few days our machines are going out of pids because of.. leaked threads.
I've spent some time to investigate it and found a way to reproduce it relatively easy (not every time, but in several minutes...).
It is just import && trim && export in an endless loop like that:
```
while [ true ]; do
    zpool import -d /root/ssd/zpool_test2/r ztest2 && zpool trim ztest2 && zpool export ztest2
done
```
The important part here is to export the pool immediately after the trim is issued. The issue occurs when the trim complete event is rised and SPA_ASYNC_MAN_TRIM_TASKQ_DESTROY is being displatched to have the spa_async_thread already suspended or not yet scheduled when it gets suspended by the export ioctl. This way the export path will not take care of destroying the man_trim_task at all and it remains out there forever.
Currently I am running totally out of tree, but I guess that should apply or at least could be easily reused:
```diff
diff --git a/module/zfs/spa.c b/module/zfs/spa.c
index 9702df0c26f3..f609d834acb1 100644
--- a/module/zfs/spa.c
+++ b/module/zfs/spa.c
@@ -4624,7 +4624,34 @@ spa_export_common(char *pool, int new_state, nvlist_t **oldconfig,
         */
        spa_open_ref(spa, FTAG);
        mutex_exit(&spa_namespace_lock);
+
+       /*
+        * Before we suspend the async thread wait if there is ongoing trim.
+        * Otherwise we may not execute the man_trim_tasq_destroy scheduled by
+        * man_trim_done because the async thread is suspended. This results in
+        * tasq thread leak. Drop the mutex after the trim completes and
+        * reacquire it later when checking for still existing taskq with async
+        * thread suspended. This way we give chance to get it regullary cleaned
+        * by the done callback async request if the asyn thread is still
+        * running. Eventually if the request havent been executed before we
+        * suspend the async_thread explixitly clean up the man_trim taskq.
+        */
+       mutex_enter(&spa->spa_man_trim_lock);
+       while (spa->spa_num_man_trimming > 0)
+               cv_wait(&spa->spa_man_trim_done_cv, &spa->spa_man_trim_lock);
+       mutex_exit(&spa->spa_man_trim_lock);
+
        spa_async_suspend(spa);
+       /*
+        * After the suspend, chech if there is still man trim taskq which is
+        * about to leak otherwise adn destroy it. This way its work will be
+        * commited by the txg sync work flushed below.
+        */
+       mutex_enter(&spa->spa_man_trim_lock);
+               if (spa->spa_man_trim_taskq)
+                       spa_man_trim_taskq_destroy(spa);
+       mutex_exit(&spa->spa_man_trim_lock);
+
        if (spa->spa_zvol_taskq) {
                zvol_remove_minors(spa, spa_name(spa), B_TRUE);
                taskq_wait(spa->spa_zvol_taskq);
```
Also I suppose the same should be done to address the auto trim ?

I have also changed the trim task que thread pool name formatting in order to get closer what is going on. That way it is much easier to identify who is who and what happens, since the task->comm is being truncated to 16 chars by default and the "_man_trim"/"_auto_trim" suffix gets truncated at all (or at least in my case). That's just a suggestion about the naming convention.
```diff
diff --git a/module/zfs/spa_misc.c b/module/zfs/spa_misc.c
index 8a91b40510da..ab0443f5537e 100644
--- a/module/zfs/spa_misc.c
+++ b/module/zfs/spa_misc.c
@@ -2257,7 +2257,7 @@ spa_auto_trim_taskq_create(spa_t *spa)
 
        ASSERT(MUTEX_HELD(&spa->spa_auto_trim_lock));
        ASSERT(spa->spa_auto_trim_taskq == NULL);
-       (void) snprintf(name, MAXPATHLEN, "%s_auto_trim", spa->spa_name);
+       (void) snprintf(name, MAXPATHLEN, "atrim_%s", spa->spa_name);
        spa->spa_auto_trim_taskq = taskq_create(name, 1, minclsyspri, 1,
            spa->spa_root_vdev->vdev_children, TASKQ_DYNAMIC);
        VERIFY(spa->spa_auto_trim_taskq != NULL);
@@ -2283,7 +2283,7 @@ spa_man_trim_taskq_create(spa_t *spa)
                 */
                return;
        }
-       (void) snprintf(name, MAXPATHLEN, "%s_man_trim", spa->spa_name);
+       (void) snprintf(name, MAXPATHLEN, "mtrim_%s", spa->spa_name);
        spa->spa_man_trim_taskq = taskq_create(name, 1, minclsyspri, 1,
            spa->spa_root_vdev->vdev_children, TASKQ_DYNAMIC);
        VERIFY(spa->spa_man_trim_taskq != NULL);
```
I am open for any critics about this fix.

Best regards,
Angel Shtilianov

**Comment by tuxoko** [comment]
@aerusso 
I'll look into it. Unfortunately, my priority have changed recently, I'll try to find time to do it.

**Comment by dweeezil** [comment]
@aerusso FYI, my post-removal-rebase branch, which I'd not yet widely publicized, is in fact based on @tuxoko's "v2" work.  I've not yet had the time to test it at all yet, short of making sure it compiles.  I hope to get it tested out in the next couple of days.

**Comment by kpande** [comment]
@dweeezil any updates?

**Comment by RJVB** [comment]
Maybe not the best place to ask, but what alternatives are there in the meantime until the change is incorporated? How about not giving the entire SSD to ZFS, but keeping a part of it for some other filesystem, or just swap - would that allow the kernel to TRIM the entire device or only the non-ZFS part?

**Comment by mailinglists35** [comment]
> 
> 
> Maybe not the best place to ask, but what alternatives are there in the meantime until the change is incorporated? How about not giving the entire SSD to ZFS, but keeping a part of it for some other filesystem, or just swap - would that allow the kernel to TRIM the entire device or only the non-ZFS part?

switch to freebsd https://blog.plein.org/2014/04/15/freebsd-9-2-supports-zfs-trimunmap/

**Comment by RJVB** [comment]
Hmm, apparently it’s possible to do this from a VM ... and I have TrueOS on a partition I can either boot from or run in a VM. I’ll just have to remember to export the pool from linux first, and maybe unmount the swap partition too.

**Comment by drescherjm** [comment]
> How about not giving the entire SSD to ZFS, but keeping a part of it for some other filesystem, or just swap - would that allow the kernel to TRIM the entire device or only the non-ZFS part?

It would be only the non zfs part

**Comment by RJVB** [comment]
>It would be only the non zfs part

Well, that's not a big deal if I can do the ZFS part by importing the pool under a FreeBSD clone. Would importing the pool be enough, or would I'd have to do a scrub or something similar?

**Comment by mailinglists35** [comment]
> 
> It would be only the non zfs part
> Well, that's not a big deal if I can do the ZFS part by importing the pool under a FreeBSD clone. Would importing the pool be enough, or would I'd have to do a scrub or something similar?

I suggest we stop discussing this here and now before a mod will lock this thread. Use the mailing lists for this.

**Comment by dweeezil** [comment]
Replaced with #8255.
